src/renderers/WebGLRenderer.js

```diff
@@ -1512,6 +1514,10 @@ function WebGLRenderer( parameters ) {
 
 				needsProgramChange = true;
 
+			} else if ( materialProperties.vertexAlpha !== vertexAlpha ) {
```

Comment:
This is exactly the problematic line I've mentioned before which affects now all calls of `setProgram()` when a material is not changed.

Reply:
How about adding a ‘layoutChanged’ flag in ‘geometry’, which gets set every time a new attribute is added or an attribute’s type/size is changed? If the user changed the attribute layout during rendering, he probably wanted the material being updated.

Reply:
How about adding a ‘layoutChanged’ flag in ‘geometry’, which gets set every time a new attribute is added or an attribute’s type/size is changed? If the user changed the attribute layout during rendering, he probably wanted the material being updated.

Reply:
It's not only the geometry that would requires these checks. You also have to check all kinds of 3D objects at this place. This can all be avoided by using material properties.

Reply:
Besides, your approach does still not solve https://github.com/mrdoob/three.js/pull/20975#discussion_r589639751.

Reply:
> It's not only the geometry that would requires these checks. You also have to check all kinds of 3D object properties. This can all be avoided by using material properties.

Had a quick look at the documentation of `Object3D` and `Objects` section. Looks like the properties affecting shader are only `geometry` and `receiveShadow`? Sorry if missed anything.

Reply:
That list is not complete, sorry.